### PR TITLE
#182631730: USERS TO HIGHLIGHT AND COMMENT ON TEXT

### PR DESCRIPTION
**Pull request by foscraft**
#### What does this Pull Request do?
+ Users can highlight, comment on any text in comment
#### Description of Task to be completed?
+ Users should be able to highlight and comment on any text in an article
#### How should this be manually tested?
+ unittesting with pytest
#### Any background context you want to provide?
+ Users should be able to highlight and comment on any text in an article
#### What are the relevant Pivotal tracker card?
+ [card](https://www.pivotaltracker.com/n/projects/2570604/stories/182631730)
#### Screenshots (if appropriate)
+  Screenshot from 2022-08-01 17-40-54
#### Questions:
N/A

### CODE REVIEW

**Review by narravabrion — COMMENTED** [review]
Code looks good

File: articles/models.py [hunk 1]
```diff
@@ -46,3 +47,31 @@ def __str__(self) -> str:
...
+        Profile, on_delete=models.CASCADE, related_name="authored"
+    )
+    article = models.ForeignKey(
+        Article, on_delete=models.CASCADE, related_name="article"
+    )
+
+    class Meta:
+        ordering = ["-createdAt"]
+
+    def __str__(self) -> str:
```
Comment (mainamuriuki274) [inline articles/models.py:72]: Won't the body be too long for an instance representation?

File: articles/models.py [hunk 2]
```diff
@@ -46,3 +47,31 @@ def __str__(self) -> str:
...
+class Comment(models.Model):
+    """
+    Handles CRUD on a comment that has been made on article
+    """
+
+    lookup_id = models.UUIDField(
+        default=uuid.uuid4, editable=False, unique=True, max_length=255
+    )
+    body = models.TextField(max_length=500, blank=False, null=False)
+    createdAt = models.DateTimeField(auto_now_add=True)
```
Comment (mainamuriuki274) [inline articles/models.py]: Use the TimeStampedModel

File: articles/models.py [hunk 3]
```diff
@@ -46,3 +47,31 @@ def __str__(self) -> str:
...
+    updatedAt = models.DateTimeField(auto_now=True)
```
Comment (mainamuriuki274) [inline articles/models.py]: This is not pythonic

File: articles/serializers.py [hunk 4]
```diff
@@ -47,3 +47,21 @@ def create(self, validated_data: Any) -> Any:
...
+            "lookup_id",
+            "body",
+            "createdAt",
+            "updatedAt",
+            "author",
+            "highlight_start",
+            "highlight_end",
+            "highlight_text",
+        )
+        read_only_fields = ("profile",)
```
Comment (mainamuriuki274) [inline articles/serializers.py]: where is thisfield?

File: articles/tests/test_models.py [hunk 5]
```diff
@@ -38,3 +42,37 @@ def test_slug_article(self) -> None:
         self.assertEqual(
             article.slug, slugify(f"{article.title}-{article.lookup_id}")
         )
+
+
+class TestCommentModel(TestCase):
+    def setUp(self) -> None:
+        adata = {
```
Comment (mainamuriuki274) [inline articles/tests/test_models.py]: give more descriptive variable names adata and pdata don't make sense

File: articles/tests/test_views.py [hunk 6]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
         )
         self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+
+class TestCommentViews(APITestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.password = fake.password()
+        cls.client = APIClient()
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:278]: Have this in setUp method

File: articles/tests/test_views.py [hunk 7]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+            "Authentication credentials were not provided.",
+        )
+
+    def test_get_all_comments(
+        self,
+    ) -> None:
+        response = self.client.get(
+            reverse("comments", kwargs={"slug": self.article.slug})
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py]: A get all comments shoulf not be  a 201,

Comment (mainamuriuki274) [inline articles/tests/test_views.py]: assert more than once

File: articles/tests/test_views.py [hunk 8]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+            reverse(
+                "comment-detail",
+                kwargs={
+                    "slug": self.article.slug,
+                    "lookup_id": self.comment.lookup_id,
+                },
+            ),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:378]: assert more than once

File: articles/tests/test_views.py [hunk 9]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+        response = self.client.get(
+            reverse(
+                "comment-detail",
+                kwargs={
+                    "slug": self.article.slug,
+                    "lookup_id": self.comment.lookup_id,
+                },
+            ),
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:400]: assert more than once

File: articles/tests/test_views.py [hunk 10]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+
+    def test_get_comment_without_authentication(
+        self,
+    ) -> None:
+        response = self.client.get(
+            reverse(
+                "comment-detail",
+                kwargs={
+                    "slug": self.article.slug,
+                    "lookup_id": self.comment.lookup_id,
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:396]: Why do you need both the lookup_id and slug for llookup? just one should should suffice

File: articles/tests/test_views.py [hunk 11]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+                },
+            ),
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_comment_with_invalid_lookup_id(
+        self,
+    ) -> None:
+        with self.assertRaises(ValidationError):
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:413]: Assert also for the message

File: articles/tests/test_views.py [hunk 12]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+                "comment-detail",
+                kwargs={
+                    "slug": self.article.slug,
+                    "lookup_id": self.comment.lookup_id,
+                },
+            ),
+            data=data,
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py]: Assert that it is updated that the body has changed
Checkout the difference between put and patch

File: articles/tests/test_views.py [hunk 13]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+            reverse(
+                "comment-detail",
+                kwargs={
+                    "slug": self.article.slug,
+                    "lookup_id": self.comment.lookup_id,
+                },
+            ),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py]: Delete status code is not 200, should be 204

File: articles/tests/test_views.py [hunk 14]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+            "Authentication credentials were not provided.",
+        )
+
+
+class TestHighlightCommentViews(APITestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.password = fake.password()
+        cls.client = APIClient()
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:597]: have this in setUp

File: articles/tests/test_views.py [hunk 15]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+        self,
+    ) -> None:
+
+        response = self.client.post(
+            reverse("comments", kwargs={"slug": self.article.slug}),
+            data=self.data,
+            **self.bearer_token,
+            format="json",
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:654]: Assert more than once

File: articles/tests/test_views.py [hunk 16]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+        login_url = reverse("login")
+        response = self.client.post(
+            login_url,
+            data={"email": self.user.email, "password": self.password},
+            format="json",
+        )
+        token = json.loads(response.content).get("access")
+        return {"HTTP_AUTHORIZATION": f"Bearer {token}"}
+
+    def test_create_highlight_comment(
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:641]: Where are the docstrings???

Comment (mainamuriuki274) [reply articles/tests/test_views.py:641]: Add them everywhere

File: articles/tests/test_views.py [hunk 17]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+    ) -> None:
+        self.data["highlight_end"] = fake.pyint(
+            min_value=len(self.article.body) + 1
+        )
+        response = self.client.post(
+            reverse("comments", kwargs={"slug": self.article.slug}),
+            data=self.data,
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py:760]: Assert for message also

File: articles/tests/test_views.py [hunk 18]
```diff
@@ -262,3 +264,307 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
...
+
+    def test_get_highlight_comments(
+        self,
+    ) -> None:
+        response = self.client.get(
+            reverse("comments", kwargs={"slug": self.article.slug})
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline articles/tests/test_views.py]: assert more than once

File: articles/views.py [hunk 19]
```diff
@@ -32,3 +35,124 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+
+    def post(self, request: Request, slug: str) -> Response:
+        """
+        Allows authenticated users can add comments on articles
+        Args:
+            slug: this a slug for a particular article
+        Returns:
+            code: The return 201 created for success
+        """
+        self.author = get_object_or_404(Profile, user=request.user)
```
Comment (mainamuriuki274) [inline articles/views.py]: I don't think it needs to have self.author or self.article  you can just have them as author and article

File: articles/views.py [hunk 20]
```diff
@@ -32,3 +35,124 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+        self.article = get_object_or_404(Article, slug=slug)
+        start_index = validate_index(request.data.get("highlight_start"), slug)
+        end_index = validate_index(request.data.get("highlight_end"), slug)
+        if start_index and end_index:
```
Comment (mainamuriuki274) [inline articles/views.py:71]: Why not have this functionality in create method in serializer

File: articles/views.py [hunk 21]
```diff
@@ -32,3 +35,124 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+        serializer.save(author=self.author, article=self.article)
+        return Response(
+            {
+                "message": "Comment added successfully",
+                "comment": serializer.data,
+            },
+            status=status.HTTP_201_CREATED,
+        )
+
+    def get(self, request: Request, slug: str) -> Response:
```
Comment (mainamuriuki274) [inline articles/views.py:90]: Most of this functionality is handled for you with the RetrieveAPIView and pagination

File: articles/views.py [hunk 22]
```diff
@@ -32,3 +35,124 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+        comments = Comment.objects.filter(article=self.article)
+        comment_count = comments.count()
+        serializer = self.serializer_class(comments, many=True)
+        return Response(
+            {
+                "message": "Comments fetched successfully",
+                "comments": serializer.data,
+                "comment_count": comment_count,
+            },
+            status=status.HTTP_201_CREATED,
```
Comment (mainamuriuki274) [inline articles/views.py]: Status code for a get is not 201

File: articles/views.py [hunk 23]
```diff
@@ -32,3 +35,124 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+        )
+
+
+class CommentDetailView(generics.GenericAPIView):
```
Comment (mainamuriuki274) [inline articles/views.py:112]: Check out the RetrieveUpdateDestroyAPIView, it has handled for you a lot of this functionality

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: articles/tests/test_views.py (hunk 10 above)
Comment (foscraft) [reply articles/tests/test_views.py:396]: Comment must belong to a specific article, hence the slug

File: articles/models.py (hunk 2 above)
Comment (MwanikiN) [inline articles/models.py]: Use the timestamp file for created_at and updated_at
Also check your variable naming, refer to PEP8 standards

File: articles/serializers.py [hunk 24]
```diff
@@ -47,3 +47,21 @@ def create(self, validated_data: Any) -> Any:
...
+
+class CommentSerializer(serializers.ModelSerializer):
+    author = ProfileSerializer(read_only=True)
+
+    class Meta:
+        model = Comment
+        fields = (
+            "lookup_id",
+            "body",
+            "createdAt",
```
Comment (MwanikiN) [inline articles/serializers.py]: Variable naming

File: articles/serializers.py [hunk 25]
```diff
@@ -47,3 +47,21 @@ def create(self, validated_data: Any) -> Any:
...
+            "updatedAt",
```
Comment (MwanikiN) [inline articles/serializers.py]: Variable naming,

**Review by MwanikiN — COMMENTED** [review]
Reviewed

**Review by MwanikiN — CHANGES_REQUESTED** [review]
remove the #in your branch

File: articles/tests/test_views.py (hunk 6 above)
Comment (foscraft) [reply articles/tests/test_views.py:278]: It is set in there

File: articles/views.py (hunk 23 above)
Comment (foscraft) [reply articles/views.py:112]: I have used it before, I was able to achieve the same with this as a learning process.

File: articles/views.py (hunk 20 above)
Comment (foscraft) [reply articles/views.py:71]: I did not see the need to since I have the validators functionality.

**Pull request closed without merging**